### DISCUSSION TITLE
Use zustand to set state after an api call (for example after a form is submitted)

### DISCUSSION THREAD

**Discussion by shellking4** [pmndrs/zustand#1274] [General] [answered]
I just discovered Zustand and really love its simplicity and ease of use.

I'm a bit beginner in React though and working on a next.js app where I need to set some global state after user authentication. 
But I'm having some issues( getting this: Error: Invalid hook call. Hooks can only be called inside of the body of a function component)
I do understand that I may be calling zustand store at a wrong place but how to do it right ?

Here is my code:

```
//content of auth.store.ts

import create from 'zustand'
import { combine } from 'zustand/middleware'

const useAuthStore = create(
    combine({
        authUser: {},
        userRoles: [],
        userIsAdmin: false,
        userIsPOLICE: false,
        userIsANAT: false,
        userIsCNSR: false,
        userIsDGI: false,
        userIsDA: false,
    }, (set) => ({
        setAuthUser: (user: any) => set(() => ({ authUser: user })),
        setUserRoles: (roles: any) => set(() => ({ userRoles: roles })),
        setUserAdmin: () => set((state) => ({ userIsAdmin: (state.userRoles as string[]).includes("Administrateur") })),
        setUserPOLICE: () => set((state) => ({ userIsPOLICE: (state.userRoles as string[]).includes("Police") })),
        setUserIsANAT: () => set((state) => ({ userIsANAT: (state.userRoles as string[]).includes("ANAT") })),
        setUserIsCNSR: () => set((state) => ({ userIsCNSR: (state.userRoles as string[]).includes("CNSR") })),
        setUserIsDGI: () => set((state) => ({ userIsDGI: (state.userRoles as string[]).includes("DGI") })),
        setUserIsDA: () => set((state) => ({ userIsDA: (state.userRoles as string[]).includes("DA") })),
        setRoles: () => set((state) => ({ 
            userIsAdmin: (state.userRoles as string[]).includes("Administrateur"),
            userIsPOLICE: (state.userRoles as string[]).includes("Police"),
            userIsANAT: (state.userRoles as string[]).includes("ANAT"),
            userIsCNSR: (state.userRoles as string[]).includes("CNSR"),
            userIsDGI: (state.userRoles as string[]).includes("DGI"),
            userIsDA: (state.userRoles as string[]).includes("DA") 
        })),
    }))
)

export default useAuthStore```

```
   //content of  LoginForm component

    const LoginForm = (props: any) => {
    const formOptions = { resolver: yupResolver(loginSchema) };
    const { register, handleSubmit, reset, formState } = useForm(formOptions);
    const { errors } = formState;
    const router = useRouter()

    const setRoles = (loginResponse: any) => {
        useAuthStore((state) => { state.setUserRoles(loginResponse.user_roles) });
        useAuthStore((state) => { state.setRoles() })
    }

    const onSubmit = async (data: any) => {
        let loader = document.getElementById("spinner");
        loader!.className = "spinner-border spinner-border-sm me-2";
        setRoles(data)
        try {
            let authResponse = await authService.loginUsingAxios(data, props.loginUrl)
            loader!.className = "";
            let loginResponse: LoginResponseSchema = authResponse.data;
            storageService.store(
                "Authorization",
                `Bearer ${loginResponse.access_token}`
            );
            storageService.store(
                "user_roles",
                loginResponse.user_roles
            );
            storageService.store(
                "user_data",
                loginResponse.user_data
            );
            let previousRoute = storageService.getPreviousPath();
            router.replace(previousRoute);
            if (previousRoute) {
            } else {
                router.replace("/");
            }
        } catch (error) {
            console.log(error)
            if ((error as any).response?.status === 400) {
                loader!.className = "";
                let loginError = "Nom d'utilisateur ou mot de passe invalide 😶️";
                let hint = "Veuillez revoir vos identifiants !";
                document.getElementById("login-error")!.innerHTML = loginError;
                Swal.fire(
                    {
                        title: loginError,
                        text: hint,
                        icon: "warning",
                        ...uiConstants.swalAnimation
                    }
                );
                reset();
                return;
            } else {
                loader!.className = "";
                let loginError = "Une erreur inconnue est survenue 😶️";
                let hint = "Veuillez réessayer !";
                document.getElementById("login-error")!.innerHTML = loginError;
                Swal.fire(
                    {
                        title: loginError,
                        text: hint,
                        icon: "warning",
                        ...uiConstants.swalAnimation
                    }
                );
                return;
            }
        }
    }

    return (
        <Fragment>
            <div className={styles["formContent"] + "page-wrapper ml-2"}>
                <div className="container-fluid" style={{ marginTop: "30px" }}>
                    <div className="row">
                        <div className="col-md-6 offset-md-2 col-12" style={{ marginRight: "30px" }}>
                            <div className="card" style={{ color: "red", background: "#fff" }}>
                                <div className="card-header" style={{ color: "white", background: "#12202A" }}>
                                    <h5 className="card-title text-center text-light">Digital Immat | Login</h5>
                                </div>
                                <div style={{ color: "#12202A" }} className="card-body px-5 py-5">
                                    <form onSubmit={handleSubmit(onSubmit)}>
                                        <div className="form-group">
                                            <label style={{ color: "#12202A" }}>Nom d'utilisateur</label>
                                            <input {...register('username')} type="text" className="form-control border-dark px-4" />
                                            <div className="errors-feedback">{errors.username?.message}</div>
                                        </div>
                                        <div className="form-group">
                                            <label style={{ color: "#12202A" }} htmlFor="">Mot de passe</label>
                                            <input {...register('password')} type="password" className="form-control border-dark" />
                                            <div className="errors-feedback">{errors.password?.message}</div>
                                        </div>
                                        <div id="login-error" style={{ textAlign: "center" }} className={styles['login-error']}></div>
                                        <div className="text-end mt-4">
                                            <button type="submit" style={{ backgroundColor: "#12202A", color: "white" }} className="btn">
                                                CONNEXION
                                                <span id="spinner" role="status"></span>
                                            </button>
                                        </div>
                                    </form>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </Fragment>
    );
}
export default LoginForm;```

Thank you for all tips so that I can go around this

**Comment by dai-shi** [comment, marked as answer]
Didn't read everything (too long 😓 ), but something like this may help:
```js
const setRoles = (loginResponse: any) => {
  useAuthStore.getState().setUserRoles(loginResponse.user_roles)
  useAuthStore.getState().setRoles()
}
```

**Reply by shellking4** [reply]
Thank you for your answer. Really helped me
